network/manager.go

```diff
@@ -692,46 +713,107 @@ func (nm *networkManager) GetEndpointID(containerID, ifName string) string {
-func cnsEndpointInfotoCNIEpInfo(endpointInfo restserver.EndpointInfo, endpointID string) *EndpointInfo {
-	epInfo := &EndpointInfo{
-		Id:                 endpointID,
-		IfIndex:            EndpointIfIndex, // Azure CNI supports only one interface
-		ContainerID:        endpointID,
-		PODName:            endpointInfo.PodName,
-		PODNameSpace:       endpointInfo.PodNamespace,
-		NetworkContainerID: endpointID,
+// saves the map of network ids to endpoints to the state file
+func (nm *networkManager) SaveState(eps []*endpoint) error {
+	nm.Lock()
+	defer nm.Unlock()
+
+	logger.Info("Saving state")
+	// If we fail half way, we'll propagate an error up which should clean everything up
+	if nm.IsStatelessCNIMode() {
+		err := nm.UpdateEndpointState(eps)
+		if err != nil {
+			return err
+		}
+	}
+
+	// we either use stateless cni and save via update endpoint state, or use the state file
+	if nm.IsStatelessCNIMode() {
+		return nil
+	}
+
+	// once endpoints and networks are in-memory, save once
+	return nm.save()
+}
+
+func (nm *networkManager) DeleteState(_ []*EndpointInfo) error {
+	nm.Lock()
+	defer nm.Unlock()
+
+	logger.Info("Deleting state")
+	// We do not use DeleteEndpointState for stateless cni because we already call it in DeleteEndpoint
+	// This function is only for saving to stateless cni or the cni statefile
+	// For stateless cni, plugin.ipamInvoker.Delete takes care of removing the state in the main Delete function
+
+	if nm.IsStatelessCNIMode() {
+		return nil
 	}
 
+	// once endpoints and networks are deleted in-memory, save once
+	return nm.save()
+}
+
+// called to convert a cns restserver EndpointInfo into a network EndpointInfo
+func cnsEndpointInfotoCNIEpInfos(endpointInfo restserver.EndpointInfo, endpointID string) []*EndpointInfo {
+	ret := []*EndpointInfo{}
+
 	for ifName, ipInfo := range endpointInfo.IfnameToIPMap {
-		// This is an special case for endpoint state that are being crated by statefull CNI
+		epInfo := &EndpointInfo{
+			EndpointID:         endpointID,      // endpoint id is always the same, but we shouldn't use it in the stateless path
+			IfIndex:            EndpointIfIndex, // Azure CNI supports only one interface
+			ContainerID:        endpointID,
+			PODName:            endpointInfo.PodName,
+			PODNameSpace:       endpointInfo.PodNamespace,
+			NetworkContainerID: endpointID,
+		}
+
+		// If we create an endpoint state with stateful cni and then swap to a stateless cni binary, ifname would not be populated
 		if ifName == "" {
 			ifName = InfraInterfaceName
 		}
-		// TODO: DelegatedNIC state will be added in a future PR
-		if ifName != InfraInterfaceName {
-			continue
-		}
+
+		// filling out the InfraNIC from the state
 		epInfo.IPAddresses = ipInfo.IPv4
 		epInfo.IPAddresses = append(epInfo.IPAddresses, ipInfo.IPv6...)
-		epInfo.IfName = ifName
+		epInfo.IfName = ifName // ifname (container veth peer) not used in linux (or even windows) deletion
```

Comment:
The comment needs to be changes since EpInfo . ifname is holding the ""eth0, et1,eth2" in the state unlike the ifname in Endpoint.

Reply:
added a comment that describes the value of epInfo.IfName in this function-- please check if it addresses your concerns